# Introduction to Python

In [1]:
input("Enter your name: ")

'Abhijit'

In [2]:
print("Hello World")

Hello World


In [3]:
!python --version #Defines the version of Python. Python is case sensitive and space sensitive

Python 3.12.2


## On numbers

#### Addition

Adding two integers:

In [ ]:
2 + 2 # Adding two integers

4

Adding a floating-point number to an integer

In [ ]:
2.0 + 2 # Adding a floating-point number to an integer

4.0

#### Subtraction

Subtraction of two integers

In [ ]:
4 - 2 # Subtraction of two integers


2

subtraction operation between an integer and a floating-point number

In [ ]:
5 - 3.0 #subtraction operation between an integer and a floating-point number

2.0

#### Multiplication

Multiplying two integers

In [ ]:
2 * 3 # Multiplying two integers

6

  Multiplying an integer by a floating-point number

In [ ]:
2 * 3.0 # Multiplying an integer by a floating-point number

6.0

Multiple operations with precedence

In [ ]:
# Performing multiple operations with precedence
2 + 3 * 5 + 3

20

In [ ]:
#Performing multiple operations with precedence
(2 + 3) * (5 + 3)

40

#### Division

Division of two integers

In [ ]:
# Division of two integers
5 / 2

2.5

Floor division of two integers

In [ ]:
# Floor division of two integers
5 // 2

2

Floor division of an integer by a floating-point number

In [ ]:
# Floor division of an integer by a floating-point number
5 // 2.0

2.0

Division of an integer by a floating-point number

In [ ]:
# Division of an integer by a floating-point number
5 / 2.0

2.5

#### Exponentiation

In [ ]:
5 ** 2 # 5 raised to the power of 2

25

In [ ]:
452 ** 0.5 # Square root of 452

21.2602916254693

Scientific notation

In [ ]:
10 + 4e2 # Scientific notation: 10 + 4e2 (equivalent to 10 + 4 * 10**2)

410.0

In [ ]:
#SyntaxError
var.4=5

SyntaxError: ignored

# On string

Concatenation of two strings

In [ ]:
# Concatenation of two strings
print("Hello " + "World")

Hello World


Slicing the first three characters from the string

In [ ]:
# Slicing the first three characters from the string
print("Hello"[:3])

Hel


Slicing from the third character till the end of the string

In [ ]:
# Slicing from the third character till the end of the string
print("Hello"[3:])

lo


# Variables

Multiple assignment to variables

In [ ]:
a = b = 10 # Multiple assignment to variables a and b
print(a)
print(a,b) # Print the values of variables 'a', 'b'
print(a+b) # Print the result of adding 'a' and 'b'

10
10 10
20


Multiple assignment using unpacking

In [ ]:
# Multiple assignment using unpacking
one, two, three = 1,"two",30.00
print(one)
print(two)
print(three)

1
two
30.0


Assigning different types to a variable

In [ ]:
# Assigning different types to a variable
number = 1
print(type(number))
str = 'string'
print(type(str))
number = str
print(number)
print(type(number))

<class 'int'>
<class 'str'>
string
<class 'str'>


Variable name can start with an underscore

In [ ]:
_varfada567=5 # Variable name can start with an underscore
print(_varfada567)

5


Variable name is case sensitive

In [ ]:
A=7
print(A) # Print the value of variable 'A', which is 7
print(a) # Print the value of variable 'a', which is still 10

7
10


Multiple statements on one line using a semicolon

In [ ]:
var4="hello";var4

'hello'

# Type of Variable & type conversion

Checking the type of the value

In [ ]:
type(2) # Checking the type of the value 2

int

In [ ]:
type(2.0) # Checking the type of the value 2.0

float

In [ ]:
type("Double quote") # Type of "Double quote"

str

In [ ]:
type(False) # Type of False

bool

In [ ]:
a = ['1','2','3'] # Creating a list 'a'
type(a)

list

In [ ]:
b = ('a','2','c') # Creating a list 'b'
type(b)

tuple

In [ ]:
c = {'a':1,'b':2,'c':3} # Creating a dictionary 'c'
type(c)

dict

In [ ]:
#If you want to specify the type of variable
var1=int(5) # var1 is an integer
var2=float(5) # var2 is a floating-point number
var3=str(5) # var3 is a string

print(var1,var2,var3)

5 5.0 5


In [ ]:
print(var1+var1) # adding the integer variable

10


In [ ]:
print(type(var1)) # Type of var1

<class 'int'>


In [ ]:
print(type(var2)) # Type of var2

<class 'float'>


In [ ]:
print(type(var1+var2)) # Type of var1 + var2 is float, because it involves a floating-point number

<class 'float'>


Convert to Float

In [ ]:
x2=30
print(float(x2)) #Convert to Float

30.0


Convert to integer

In [ ]:
x3=1.567
print(int(x3)) #convert to integer

1


Rounding to the nearest integer

In [ ]:
print(type(round(x3)))#rounds to the nearest integer

<class 'int'>


Converts integer to string

In [4]:
x4=65
x5=str(x4) # Converts integer to string
print(type(x4)) # type indicates the variable type
print(type(x5))

<class 'int'>
<class 'str'>


# Commenting

In [ ]:
#This is used to write a comment

In [ ]:
#Multi Line comments
'''This is Madhu
I will be your mentor
For this course
'''

'This is Madhu\nI will be your mentor\nFor this course\n'

# Operators

Handling SyntaxError

In [ ]:
#SyntaxError: unexpected EOF while parsing
x=8
y=3
print(x     #SyntaxError
var.4=5

#SyntaxError
This is used to write a comment-y


SyntaxError: ignored

Modulus operator

In [ ]:
#Modulus operator
x=5
y=2
print(x%y)

1


ZeroDivisionError

In [ ]:
#ZeroDivisionError
x = 5
y = 2
result = x / 0
print(result)


ZeroDivisionError: ignored

Assignment operators

In [ ]:
#Assignment operators
var5=32
#var5 +=10 . The output will be 32+10
var5 +=10
print(var5)

42


In [ ]:
var5 -=20 #This is equal to var5=var5-20
print(var5)

22


In [ ]:
var5 *=3 #This is equal to var5=var5*3
print(var5)

66


In [ ]:
var6=var5
print(var6)


198


Absolute value of a number

In [ ]:
x=-25
print(x)
print(abs(x)) # abs() returns the absolute value of a number

-25
25


In [ ]:
#Comparison operator
#==  Equal to
#>   Greater than operator
#>=  Greater than or equal to
#< <= != ( not equal to )

'''
#Logical Operators
and
or
Not
'''

#On Binary

Convert to binary

In [ ]:
x1=bin(8) #convert to binary. 000=0,001 =1, 010=2, 011=3
print(x1)

0b1000


Logical 'and' operation

In [ ]:
x=3
y=8
print(x and y)  # Prints the result of the logical 'and' operation between x and y

8


Logical 'or' operation

In [ ]:
print(x or y)  # Prints the result of the logical 'or' operation between x and y

3


Bitwise AND

In [ ]:
print(x & y) # Prints the result of bitwise AND between x and y

0


Bitwise OR

In [ ]:
print(2 | 3) # Prints the result of bitwise OR between 2 and 3

3


In [ ]:
#Binary value of 2 is 10 . Binary value of 3 is 11 . so 10&11=10 (decimal 2)

In [ ]:


# Left Shift
x = 5   # Binary: 0101
shifted_left = x << 2
print(shifted_left)

# Right Shift
x = 16  # Binary: 10000
shifted_right = x >> 2
print(shifted_right)


4
14
20
4


#ASCII

In [ ]:
char = '^'
ascii_value = ord(char)
print(ascii_value)


94


In [ ]:
binary_value = '10000000'
decimal_value = int(binary_value, 2)
print(decimal_value)

In [ ]:
characters = ['z', 'B', 'C', 'D']

# Iterate over each character
for char in characters:
    ascii_value = ord(char)
    binary_value = bin(ascii_value)[2:]  # Remove the '0b' prefix
    print(f"Binary representation of '{char}' is {binary_value}")


Binary representation of 'z' is 1111010
Binary representation of 'B' is 1000010
Binary representation of 'C' is 1000011
Binary representation of 'D' is 1000100


In [ ]:
sentence = "Hello, Sony!"

# Iterate over each character in the sentence
for char in sentence:
    ascii_value = ord(char)
    binary_value = bin(ascii_value)
    print(f"Binary representation of '{char}' is {binary_value}")


Binary representation of 'H' is 0b1001000
Binary representation of 'e' is 0b1100101
Binary representation of 'l' is 0b1101100
Binary representation of 'l' is 0b1101100
Binary representation of 'o' is 0b1101111
Binary representation of ',' is 0b101100
Binary representation of ' ' is 0b100000
Binary representation of 'S' is 0b1010011
Binary representation of 'o' is 0b1101111
Binary representation of 'n' is 0b1101110
Binary representation of 'y' is 0b1111001
Binary representation of '!' is 0b100001


The binary representations for 'A', 'B', and 'C' are 01000001, 01000010, and 01000011, respectively. Each binary representation consists of 8 bits, with '0' representing a bit value of 0 and '1' representing a bit value of 1.

# Let us work with strings

Convert string to upper case

In [ ]:
string1='upper_case'
print(string1.upper()) # convert string to upper case

UPPER_CASE


Convert string to lowercase

In [ ]:
string2='ScieNce'
print(string2.lower()) # Convert string to lowercase

science


Concatenate two strings

In [ ]:
first_name="steve"
last_name='jobs'
full_name=first_name + ' '+last_name  # Concatenate two strings
print(full_name.title())

Steve Jobs


TypeError - mixing string and integer types in an operation

In [ ]:
#TypeError
greeting = "Hello"
age = 25
message = greeting + age
print(message)

TypeError: can only concatenate str (not "int") to str

 Find the first occurrence of word in the string

In [ ]:
string15="I am teaching python prerequisites to working professionals of industry"
location_Data=string15.find('python')
print(location_Data)

14


Find the first occurrence of letter in the string

In [ ]:
string15="I am teaching python prerequisites for industry"
location_Data=string15.find('a')
print(location_Data)

2


Replacing substring

In [ ]:
#Replacing substring
string16=string15.replace('industry','Sony')
print(string16)

I am teaching python prerequisites for Sony


Splitting the string into a list of words

In [ ]:
string17=string16.split(' ') #split command splits the string based on keyword
print(string17)

['I', 'am', 'teaching', 'python', 'prerequisites', 'for', 'Sony']


# White Space refers to to characters that is not visible to us. Examples are tab (\t) newline(\n)

In [ ]:
print('Hello World')

Hello World


Tab character before the string

In [ ]:
print("\tHello World") # Tab character before the string

	Hello World


In [ ]:
#SyntaxError
print(\t"Hello World")

SyntaxError: ignored

Tab character between words

In [ ]:
print("Hello\t World") # Tab character between 'Hello' and 'World'

Hello	 World


Newline character between words

In [ ]:
print("Hello \nWorld") # Newline character between 'Hello' and 'World'

Hello 
World


Multiple newline characters

In [ ]:
print("Hello \n\n\n\n\n World") # Multiple newline characters

Hello 




 World


# Remove white space using lstrip / rstrip and strip

In [ ]:
x="  Hello World   "
print(x)

  Hello World   


strip white spaces on left side

In [ ]:
print(x.lstrip()) #strip white spaces on left side

Hello World   


In [ ]:
print(x + 'v')

  Hello World   v


In [ ]:
print(x.rstrip()+ 'v')

  Hello Worldv


strip white spaces on the right side

In [ ]:
print('v'+x+'v') # strip white spaces on the right side

v  Hello World   v


In [ ]:
print('v'+x.strip()+'v')

vHello Worldv


In [ ]:
x

'  Hello World   '

In [ ]:
#AttributeError
print('v' + x.strip() + 'v')

vHello Worldv


# List

List are collection of objects and can contain string, integer, float

Create lists

In [ ]:
#List
ex1=[4,5,6,10]
print(ex1)

[4, 5, 6, 10]


In [ ]:
ex2=['python','java', 'php']
print(ex2)

['python', 'java', 'php']


In [ ]:
ex3=['python',4,10.2]

Access the first element of list

In [ ]:
print(ex3)
print(ex3[0])

['python', 4, 10.2]
python


In [ ]:
#List stores data in ordered way, we can change values of the list elements, I can create duplicates
ex3[2]=25 #change 10.2 to 25
print(ex3)

['python', 4, 25]


Negative Indexing

In [ ]:
#Negative Indexing
print(ex3[-1])

100


Slicing

In [ ]:
ex4=[4,8,9,2,3,4,20]
print(ex4[2:5]) # Slicing elements from index 2 to 4

[9, 2, 3]


In [ ]:
print(ex4[2:]) # Slicing elements from index 2 to the end

[9, 2, 3, 4, 20]


In [ ]:
print(ex4[:2]) # Slicing elements from the beginning to index 1

[4, 8]


In [ ]:
 print(ex4[-4]) # Access the fourth element from the end

2


In [ ]:
print(ex4)

[4, 8, 9, 2, 3, 4, 20]


Insert the number at the end of list

In [ ]:
ex4.append(25) #Insert the number at the end of list
print(ex4)

[4, 8, 9, 2, 3, 4, 20, 25]


Insert an item in the list for the given index number

In [ ]:
ex4.insert(4,100) # insert an item in the list for the given index number
print(ex4)

[4, 8, 9, 2, 100, 3, 4, 20, 25]


In [ ]:
names=['Raj','Arun','Madhu','Suresh']

In [ ]:
#lists are indexed and starts from 0
print(names[0])
print(names[3])

Raj
Suresh


In [ ]:
#Lists can also be accessed using negative numbers where -1 represents the last value in the list
print(names[-1]) #printing the last value in the list
print(names[-3]) #printing the 3rd last value from the list

Suresh
Arun


In [ ]:
name1=names

In [ ]:
print(name1)

['Raj', 'Kumar', 'Madhu', 'Suresh']


Change the element in a list

In [ ]:
name1[1]='Kumar' # Change the element at index 1 in the list to 'Kumar'
print(name1)

['Raj', 'Kumar', 'Madhu', 'Suresh']


Finding the location of an item in the list

In [ ]:
#finding the location of an item in the list
print(name1.index('Madhu')) # Find the index of 'Madhu' in the list

2


In [ ]:
#print(name1.index('madhu')) #madhu is not in the list

Check if an item exists in the list

In [ ]:
print('Madhu' in name1)
print('madhu' in name1)


True
False


In [ ]:
name1.append('Ganesh')#appends to end of list
print(name1)

['Raj', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']


Insert an item at a specific location

In [ ]:
#insert item at specific location
#Let me add Arvind after Raj
name1.insert(1,'Arvind')
print(name1)

['Raj', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']


In [ ]:
name2=[] # Created an empty list
name2.append('George') # Append 'George' to the end of the list
print(name2)

['George']


In [ ]:
name2.insert(0,'Mary') # Insert 'Mary' at index 0 in the list
print(name2)

['Mary', 'George']


SORTED AND JOIN

In [ ]:
name3='Madhu'
name4=sorted(name3) # Sort the characters of the string 'Madhu' and create a list
print(name4)
print(name4)

['M', 'a', 'd', 'h', 'u']


In [ ]:
name5='*'.join(name4) # Join the characters in the list with '*' separator to form a string
print(name5)

M*a*d*h*u


In [ ]:
print(name3) # The original string 'Madhu' remains unchanged

Madhu


In [ ]:
name6=['x','Y','a', 'A']
for i in sorted(name6):
    print(i)  # Print the sorted elements of the list 'name6'

A
Y
a
x


In [ ]:
name6.sort()# Sort the list 'name6' in-place (modifies the list itself)

In [ ]:
print(name6)

['A', 'Y', 'a', 'x']


In [ ]:
name7=['ak','ab','aa']
name7.sort()
print(name7)

['aa', 'ab', 'ak']


In [ ]:
name7.sort(reverse=True) #Sort in descending order

In [ ]:
print(name7)

['ak', 'ab', 'aa']


sort() will directly sort the list and does not return a value (none). sorted() will not change the list but returns the sorted value

In [ ]:
print(name1)

['Raj', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']


In [ ]:
name1.reverse() # reverses the list
print(name1)

['Ganesh', 'Suresh', 'Madhu', 'Kumar', 'Arvind', 'Raj']


In [ ]:
print(name2)

['Mary', 'George']


In [ ]:
#clear all items in the list
name2.clear()
print(name2)

[]


In [ ]:
print(name1)

['Ganesh', 'Suresh', 'Madhu', 'Kumar', 'Arvind']


In [ ]:
name1.pop(5) #removes or pops the given element from the list
print(name1)

['Ganesh', 'Suresh', 'Madhu', 'Kumar', 'Arvind']


In [ ]:
#Index error
name1.pop(5) #removes or pops the given element from the list
print(name1)

IndexError: pop index out of range

In [ ]:
name1.pop() #to remove the last element in the list
print(name1)

['Ganesh', 'Suresh', 'Madhu', 'Kumar']


In [ ]:
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']

In [ ]:
#remove an item by value
name1.remove('Madhu')
print(name1) #only the first instance of Madhu is removed. If there are two elements of Madhu, then the first is removed

['Raj', 'Arvind', 'Arvind', 'Kumar', 'Suresh', 'Ganesh']


In [ ]:
#Remove an item using delete command
del name1[2] # removing the second Arvind
print(name1)

['Raj', 'Arvind', 'Kumar', 'Suresh', 'Ganesh']


In [ ]:
#delete multiple items
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']
del name1[:3] #delete the first 3 list values
print(name1)

['Kumar', 'Madhu', 'Suresh', 'Ganesh']


In [ ]:
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']
del name1[3:] # Delete elements from index 3 to the end of the list
print(name1)

['Raj', 'Arvind', 'Arvind']


In [ ]:
#find the length of a list
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']
print(len(name1))

7


In [ ]:
#Slicing a list
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']
name2=name1[2:5] #Note that Suresh will not be printed
print(name2)

['Arvind', 'Kumar', 'Madhu']


In [ ]:
name1=['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']
#copy an entire list
copied_name1=name1[:]
print(copied_name1)

['Raj', 'Arvind', 'Arvind', 'Kumar', 'Madhu', 'Suresh', 'Ganesh']


In [ ]:
copied_name2=name1[2:5]
print(copied_name2)

['Arvind', 'Kumar', 'Madhu']


In [ ]:
t3 = ['a','b','c','d']
'a' in t3 # Check if 'a' exists in the list t3

True

# Tuples

Tuples are lists whose values cannot be changed. We cannot use 'insert', 'append' commands in tuples. However I can remove items

In [ ]:
vegetable=('Brinjal','Onion','Tomato')
print(vegetable[0]) # Access the first element of vegetable, which is 'Brinjal'
print(type(vegetable)) # Print the type of the variable 'vegetable', which is tuple

Brinjal
<class 'tuple'>


In [ ]:
#We will get error if we try to append
vegetable.append('Potato')

AttributeError: 'tuple' object has no attribute 'append'

In [ ]:
# Tuples can contain mixed object types
t2 = (1,'two',['three','four','five'],{'key':'value'})
print(type(t2))
t2

<class 'tuple'>


(1, 'two', ['three', 'four', 'five'], {'key': 'value'})

In [ ]:
t2[0] # Access the first element of t2

1

In [ ]:
t2[2][2] # Access the third element of the list within t2

'five'

In [ ]:
t2[3]['key'] # Access the value of the dictionary within t

{'key': 'value'}

In [ ]:
t3 = (1)
t3

1

In [ ]:
t3 = (1,2,3)
t4 = ('a','b')
t3 + t4 # Concatenate t3 and t4

(1, 2, 3, 'a', 'b')

In [ ]:
t3 * 5

(1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3)

 Immutability

In [ ]:
# tuples are immutable

t2[0]=2 # this will create an error
t2

TypeError: 'tuple' object does not support item assignment

In [ ]:
# although tuples are immutable if the object in the tuple is mutable then data can be changed
t2[3]['key']='no value'
t2

(1, 'two', ['three', 'four', 'five'], {'key': 'no value'})

In [ ]:
t2[2].append('six')
t2

(1, 'two', ['three', 'four', 'five', 'six'], {'key': 'no value'})

 Indexing and Slicing in tuple

In [ ]:
t2[0:2]

(1, 'two')

In [ ]:
t2[-1]

{'key': 'no value'}

In [ ]:
len(t2)

4

In [ ]:
t2[::-1]

({'key': 'no value'}, ['three', 'four', 'five', 'six'], 'two', 1)

 Tuples Methods

In [ ]:
# index gives the position of the element
t2.index(1)

0

In [ ]:
# count of elements present in the tuple
t2.count(1)

1

In [ ]:
# cannot delete tuple element
print(t2)
del(t2[2]) # this will create error
t2

(1, 'two', ['three', 'four', 'five', 'six'], {'key': 'no value'})


TypeError: 'tuple' object doesn't support item deletion

In [ ]:
t2
del(t2)
t2 # this will create error

NameError: name 't2' is not defined

 Tuple Membership Test

In [ ]:
t3 = ('a','b','c','d')
'a' in t3

True

In [ ]:
'z' in t3

False

Others

In [ ]:
vegetable1=['Brinjal','Onion','Tomato']
print("I have purchased %s,  %s and %s" %(vegetable1[0],vegetable1[1],vegetable1[2]))

I have purchased Brinjal,  Onion and Tomato


In [ ]:
x=15
print("I have INR.%s" %x)

I have INR.15


In [ ]:
print("I have INR. %d" %x)

I have INR. 15


In [ ]:
#print("I have purchased %d,  %d and %d" %(vegetable1[0],vegetable1[1],vegetable1[2]))

In [ ]:
number1=[2,4,6]
print('I like the numbers ' + str(number1[0]) + ', '+ str(number1[1]) + ' and ' + str(number1[2]))

I like the numbers 2, 4 and 6


# THE RANGE FUNCTION


In [ ]:
x5=range(1,11)
print(x5)

range(1, 11)


In [ ]:
new_list=list(range(2,10)) #creating a list using list command
print(new_list)

[2, 3, 4, 5, 6, 7, 8, 9]


In [ ]:
print(min(new_list)) #smallest number in the list

2


In [ ]:
print(max(new_list)) #largest number in the list

9


In [ ]:
print(sum(new_list)) # sums the elements in the list

44


In [ ]:
#SyntaxError
list(range(2,10)

SyntaxError: unexpected EOF while parsing (770437954.py, line 2)

GOLDEN RULES FOR PYTHON PROGRAMMING
indentation - Use 4 spaces for indentation
line length - About 70 characters
blank lines - Use blank lines to break your code into meaningful blocks
Variable naming - use lowercase, underscore and numbers

# Functions

FUNCTIONS
Reuse of code. Do not repeat
Syntax of a function

def function_name(argument1,argument2):
    write the repetitive code that is going to be reused
    
Call the function using function_name

In [ ]:
print('Thank you Shruti for being attentive in class')
print("You will learn a lot about Python")

print('Thank you Kumar for being attentive in class')
print("You will learn a lot about Python")

print('Thank you Suresh for being attentive in class')
print("You will learn a lot about Python")


Thank you Shruti for being attentive in class
You will learn a lot about Python
Thank you Kumar for being attentive in class
You will learn a lot about Python
Thank you Suresh for being attentive in class
You will learn a lot about Python


In [ ]:
def attention_class(name):
    print("\nThank you %s for being attentive in class" %name)
    print('\nYou will learn a lot about Python')

In [ ]:
attention_class('Shruti')
attention_class('Kumar')
attention_class('Suresh')



Thank you Shruti for being attentive in class

You will learn a lot about Python

Thank you Kumar for being attentive in class

You will learn a lot about Python

Thank you Suresh for being attentive in class

You will learn a lot about Python


In [ ]:
def cubic(x):
    y= x*x*x
    return y #return is used to return a value to the outside world




In [ ]:
cubic(2)

8

In [ ]:
print(cubic(3))


27


In [ ]:
def attention_class(name):
    print("\nThank you %s for being attentive in class" %name)
    print('\nYou will learn a lot about python')

In [ ]:
attention_class('shruti')


Thank you shruti for being attentive in class

You will learn a lot about python


IF STATEMENT

if condition:
    Statement to execute if condition is true
    
if condition:
    statement 1
    statement 2

statement 3
    
    
    
    
 if-else condition:

 if(condition)
     Execute this block of code if condition = True

else:
    Execute this block of code if condition = False

In [ ]:
i=10
if(i>15):
    print("15 is less than 20")
print('This print statement is outside the if condition')

This print statement is outside the if condition


In [ ]:
i=10
if(i>15):
    print("15 is less than 20")
else:
    print('The input number is smaller than 15')

The input number is smaller than 15


Nested if

if(condition1):
    Execute this block when condition 1 is true
    if(condition2):
        Execute this block if condition 2 is true
    The inner if block ends here
The outer if block ends here

In [ ]:
i=21
if(i>10):
    if(i<15):
        print( 'i is smaller than 15')

    if(i>18):
        print('i is larger than 18')
    else:
        print('i is smaller than 18')


i is larger than 18


#if elif else

if(condition):
  execute this if true
elif:
   check the condition here
elif:
    Check the condition here
else:
    execute this statement

In [ ]:
i=25
if(i==10):
    print('i is 10')
elif(i==15):
    print('i is 15')
elif(i==20):
    print('i is 20')
else:
     print('None of the above')

None of the above


In [ ]:
#short form of if-else
i=10
print('True') if i<15 else print('False')

True


# Loop

For loop to print numbers from 1 to 10

In [ ]:
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


For loop to print numbers from 1 to 14 with a step of 3

In [ ]:
for i in range(1,15,3):  #I am specifying the step for range function
    print(i)

1
4
7
10
13


For loop to iterate over elements in a list

In [ ]:
#For statement is used for iteration
list1=['hello','Madhu','tutor','Sony']
for i in list1:
    print(i)

hello
Madhu
tutor
Sony


For loop to iterate over elements in a tuple

In [ ]:
tuple1=('hello','Madhu','tutor','V')
for i in tuple1:
    print(i)

hello
Madhu
tutor
V


For loop to iterate over characters in a string

In [ ]:
string1='Madhu'
for i in string1:
    print(i)

M
a
d
h
u


Continue, Break, and Pass statements

The continue statement skips the current iteration and continues to the next one

In [ ]:
#Continue, Break and pass statements
string1='Madhua'
for i in string1:
    if i=='a' or i=='M':
        continue
    print(i)

d
h
u


The break statement terminates the loop prematurely

In [ ]:
string1='Madhua'
for i in string1:
    if i=='h' or i=='f':
        break
    print(i)

M
a
d


The pass statement is a null operation and does nothing

In [ ]:
string1='Madhua'
for i in string1:
    pass
print(i)

a


Using range to create a list and calculating its sum

In [ ]:
x=range(1,5)
x=list(x)
print(x)

[1, 2, 3, 4]


In [ ]:
print(sum(x))

10


Calculating the sum of numbers from 1 to 4 using a for loop

In [ ]:
output=0
for i in range(1,5):
    output=output+i
print(output)

10


Looping through a list using the index

In [ ]:
list1=['Banana','Orange','Pine Apple']
for i in range(len(list1)):
    print(list1[i])

Banana
Orange
Pine Apple


Looping through a list directly

In [ ]:
names=['Raj','Arun','Madhu','Suresh']

In [ ]:
for i in names: #Looping through each element in a list
    print("My friend name is :",i) #indentation is compulsory and states that 'print' is part of 'for' code block

My friend name is : Raj
My friend name is : Arun
My friend name is : Madhu
My friend name is : Suresh


In [ ]:
for i in names: #Looping through each element in a list
    print("My friend name is :",i)
print("Hello") #Here 'print' is not part of for code block

My friend name is : Raj
My friend name is : Arun
My friend name is : Madhu
My friend name is : Suresh
Hello


In [ ]:
for i in names: #Looping through each element in a list
    print("My friend name is :",i)
    print("Hello")

My friend name is : Raj
Hello
My friend name is : Arun
Hello
My friend name is : Madhu
Hello
My friend name is : Suresh
Hello


While loop to execute a block of statements repeatedly until the condition is satisfied


In [ ]:
#While loop is used to execute a block of statements repeatedly until the given condition is satisfied
count=0
while(count<=5):
    print('Sony')
    count=count+1

Sony
Sony
Sony
Sony
Sony
Sony


Single statement while block

In [ ]:
#single statement while block
count=0
while(count<5):count +=1; print('Hello Madhu')

Hello Madhu
Hello Madhu
Hello Madhu
Hello Madhu
Hello Madhu


Continue statement in a while loop

In [ ]:
#Let us check the continue statement in while loop
i=0
a='Sony'
while i<len(a):
    if a[i]=='i' or a[i]=='l':
        i=i+1
        continue
    print(a[i])
    i=i+1

S
o
n
y


Nested while loop with an outer and inner loop

In [ ]:
a = int(input('Enter the number of loop iterations: '))
outer_counter = 0

while outer_counter < int(a):
    inner_counter = 0

    while inner_counter < 3:
        print("Outer loop:", outer_counter, "Inner loop:", inner_counter)
        inner_counter += 1

    outer_counter += 1


Enter the number of loop iterations: 20
Outer loop: 0 Inner loop: 0
Outer loop: 0 Inner loop: 1
Outer loop: 0 Inner loop: 2
Outer loop: 1 Inner loop: 0
Outer loop: 1 Inner loop: 1
Outer loop: 1 Inner loop: 2
Outer loop: 2 Inner loop: 0
Outer loop: 2 Inner loop: 1
Outer loop: 2 Inner loop: 2
Outer loop: 3 Inner loop: 0
Outer loop: 3 Inner loop: 1
Outer loop: 3 Inner loop: 2
Outer loop: 4 Inner loop: 0
Outer loop: 4 Inner loop: 1
Outer loop: 4 Inner loop: 2
Outer loop: 5 Inner loop: 0
Outer loop: 5 Inner loop: 1
Outer loop: 5 Inner loop: 2
Outer loop: 6 Inner loop: 0
Outer loop: 6 Inner loop: 1
Outer loop: 6 Inner loop: 2
Outer loop: 7 Inner loop: 0
Outer loop: 7 Inner loop: 1
Outer loop: 7 Inner loop: 2
Outer loop: 8 Inner loop: 0
Outer loop: 8 Inner loop: 1
Outer loop: 8 Inner loop: 2
Outer loop: 9 Inner loop: 0
Outer loop: 9 Inner loop: 1
Outer loop: 9 Inner loop: 2
Outer loop: 10 Inner loop: 0
Outer loop: 10 Inner loop: 1
Outer loop: 10 Inner loop: 2
Outer loop: 11 Inner loop: 0
Oute

# Sets

A set is an unordered collection with no duplicate elements. Basic uses include membership testing and eliminating duplicate entries. Set objects also support mathematical operations like union, intersection, difference, and symmetric difference.

### Creating Sets

In [ ]:
set1 = set('Moment of Truth')
set1

{' ', 'M', 'T', 'e', 'f', 'h', 'm', 'n', 'o', 'r', 't', 'u'}

In [ ]:
set2 = set()  # Creates an empty set
set2.add('A')  # Adds 'A' to set2
set2

{'A'}

In [ ]:
# set is similar to dictionary but containing only keys
setA = {'Apple','America','August'}
setA

{'America', 'Apple', 'August'}

Set can contain only unique entries. If you try to duplicate an entry in the set

In [ ]:
set3 = set()
set3.add('Blue')
set3.add('Green')
print(set3)
set3.add('Blue')
print(set3)

{'Blue', 'Green'}
{'Blue', 'Green'}


In [ ]:
# please note set is case-sensitive
colors = ['Blue','Green','Red','White','blue','Blue','White','Black']
unique_colors = set(colors)
print(unique_colors)

{'Red', 'Blue', 'Black', 'Green', 'blue', 'White'}


Sets will not allow any mutable objects. As you can see tuples are accepted as elements in a set while list is not which is a mutable object.

In [ ]:
set4 = {'A',('B','C')}
set4

{('B', 'C'), 'A'}

In [ ]:
set5 = {'X',['Y','Z']} # will generate error
set5

TypeError: unhashable type: 'list'

### Mutable and Immutable Sets

Although some of the mutable elements (like list) are not allowed in set, set itself is mutable.

In [ ]:
# Mutable Set
set6 = set(['A','B','C','D','E'])
set6.add('F')
print(set6)

{'E', 'D', 'C', 'F', 'A', 'B'}


In [ ]:
# Immutable Set
set7 = frozenset(['A','B','C'])
set7.add('D') # will generate error
set7

AttributeError: 'frozenset' object has no attribute 'add'

In [ ]:
# Sets are unordered, to get ordered set use sorted function

set8 = sorted(set6)
set8

['A', 'B', 'C', 'D', 'E', 'F']

Set Operations

Add, Remove Elements

In [ ]:
set6.clear()
set6

set()

In [ ]:
print(set4)
set4.remove('A') # will report an error if element is not present
print(set4)
set4.discard('A') # will not report an error if element is not present
print(set4)
set4.remove('A') # this will create an error

NameError: ignored

In [ ]:
print(len(set4))

In [ ]:
set4.add('A')
set4.add(('D','E'))
print(set4)

In [ ]:
# add will not take any mutable object like list
set4.add(['1','2'])
set4 # this will create an error

In [ ]:
# Union Operator
colorsA = {'Blue','Black','White','Red','Green'}
colorsB = {'Blue','White'}
colorsA.union(colorsB)

{'Black', 'Blue', 'Green', 'Red', 'White'}

In [ ]:
# Intersection operator
colorsA.intersection(colorsB)

{'Blue', 'White'}

In [ ]:
# Create a set called 'x' with elements 'a', 'b', 'c', 'd', and 'e'.
x = {"a", "b", "c", "d", "e"}

# Remove an arbitrary element from the set 'x' using the 'pop()' method.
x.pop()

'a'

In [ ]:
# Remove another arbitrary element from the set 'x'.
x.pop()

'e'

In [ ]:
# Difference between sets
x = {"a","b","c","d","e"}
y = {"b","c"}
z = {"c","d"}
x.difference(y)

{'a', 'd', 'e'}

In [ ]:
# Find the elements in 'x' that are not in 'y' using the 'difference()' method.
x.difference(y).difference(z)

{'a', 'e'}

In [ ]:
# instead of using the method difference, we can use the operator "-"
print(x - y)
print(x - y - z)

{'a', 'e', 'd'}
{'a', 'e'}


In [ ]:
# difference_update removes all elements of another set from this set.
x = {"a","b","c","d","e"}
y = {"b","c"}
x.difference_update(y)
print(x)

{'a', 'e', 'd'}


In [ ]:
# x.difference_update(y) is the same as "x = x - y"
x = {"a","b","c","d","e"}
y = {"b","c"}
x = x - y
print(x)

{'a', 'e', 'd'}


# Dictionary

Dictionary is an unordered set of key-value pairs where the keys are unique. Dictionaries are indexed by keys, which can be any immutable type; strings and numbers can always be keys. Tuples can be used as keys if they contain only strings, numbers, or tuples.

Create a dictionary with key-value pairs.

In [ ]:
dict1 = {'id': 1, 'name':'John Doe', 'email':'john.doe@example.org','salary':14000.00}
dict1

{'id': 1,
 'name': 'John Doe',
 'email': 'john.doe@example.org',
 'salary': 14000.0}

Access the value associated with the key

In [ ]:
dict1['name']

'John Doe'

Change the value associated with the key

In [ ]:
dict1['email']='john.doe@example.com'

In [ ]:
dict1

{'id': 1,
 'name': 'John Doe',
 'email': 'john.doe@example.com',
 'salary': 14000.0}

Type of Dictionary

In [ ]:
type(dict1)

dict

Convert dictionary to a string.

In [ ]:
str(dict1)

"{'id': 1, 'name': 'John Doe', 'email': 'john.doe@example.com', 'salary': 14000.0}"

 Get a list of keys

In [ ]:
list(dict1.keys())

['id', 'name', 'email', 'salary']

Get a list of values

In [ ]:
list(dict1.values())

[1, 'John Doe', 'john.doe@example.com', 14000.0]

Get the sorted list of keys

In [ ]:
sorted(list(dict1.keys()))

['email', 'id', 'name', 'salary']

In [ ]:
del(list) # this will create error

NameError: name 'list' is not defined

In [ ]:
dict2 = dict([('id',2),('name','Jack Jill'),('salary',15500)])
dict2

{'id': 2, 'name': 'Jack Jill', 'salary': 15500}

Create a dictionary 'dict3' using keyword arguments

In [ ]:
dict(id=3, salary=25000)

{'id': 3, 'salary': 25000}

In [ ]:
dict = [dict1,dict2] # Create a list containing 'dict1' and 'dict2'.
dict

[{'id': 1,
  'name': 'John Doe',
  'email': 'john.doe@example.com',
  'salary': 14000.0},
 {'id': 2, 'name': 'Jack Jill', 'salary': 15500}]

In [ ]:
dict[1]['name'] # Access the value associated with the key 'name' in 'dict2' using the list of dictionaries.

'Jack Jill'

### Dictionary Comprehension

Create a dictionary with squares of even numbers from 0 to 9.


In [ ]:
{x: x**2 for x in range(10) if x%2 == 0}

{0: 0, 2: 4, 4: 16, 6: 36, 8: 64}